### ISSUE TITLE
# feat: Implement HeadLossCalculator Protocol

### ISSUE THREAD

**Issue by ccirone2** [ccirone2/opensolve-pipe#124]
# feat: Implement HeadLossCalculator Protocol

## Overview

Implement the `HeadLossCalculator` Protocol to provide a unified interface for components that cause head loss in the hydraulic network.

## Why

**Current pain point:** Head loss calculations are scattered across the codebase with different approaches per component type:

- **Pipes:** `friction.py:calculate_pipe_head_loss_fps()` - Darcy-Weisbach
- **Valves:** Inline in solver - Cv-based or K-factor
- **Heat exchangers:** `pressure_drop * (flow/design_flow)²` - quadratic scaling
- **Strainers:** K-factor or fixed pressure drop
- **Orifices:** Discharge coefficient calculation

The solver has to know the internals of each component type to calculate losses.

**With Protocol:** Each component owns its head loss calculation. The solver just calls `calculate_head_loss()` on any component that implements the protocol.

```python
# Before (in solver)
if isinstance(comp, ValveComponent):
    if comp.cv:
        h_loss = calculate_valve_loss_cv(flow, comp.cv, sg)
    else:
        h_loss = calculate_valve_loss_k(flow, comp.valve_type, diameter)
elif isinstance(comp, HeatExchanger):
    h_loss = comp.pressure_drop * (flow / comp.design_flow) ** 2
elif isinstance(comp, Strainer):
    if comp.k_factor:
        h_loss = comp.k_factor * velocity_head
    else:
        h_loss = comp.pressure_drop * (flow / comp.design_flow) ** 2
# ... more branches

# After
if isinstance(comp, HeadLossCalculator):
    h_loss = comp.calculate_head_loss(flow_gpm, velocity_fps, fluid_props)
```

## Protocol Definition

```python
# protocols/components.py
from typing import Protocol, TYPE_CHECKING

if TYPE_CHECKING:
    from ..models.fluids import FluidProperties

class HeadLossCalculator(Protocol):
    """Components that cause head loss in the hydraulic network.

    Implementors must provide a method to calculate head loss given
    flow conditions. The calculation method varies by component type:

    - Valves: Cv-based or K-factor based
    - Heat Exchangers: Quadratic scaling from design conditions
    - Strainers: K-factor or fixed pressure drop
    - Orifices: Discharge coefficient calculation

    All head loss values are returned in feet of fluid.
    """

    def calculate_head_loss(
        self,
        flow_gpm: float,
        velocity_fps: float,
        fluid_props: "FluidProperties",
    ) -> float:
        """Calculate head loss for given flow conditions.

        Args:
            flow_gpm: Volumetric flow rate in gallons per minute
            velocity_fps: Flow velocity in feet per second
            fluid_props: Fluid properties (density, viscosity, etc.)

        Returns:
            Head loss in feet of fluid
        """
        ...
```

### 1. Add Protocol to `protocols/components.py`

Add the `HeadLossCalculator` Protocol as shown above.

### 2. Implement for `ValveComponent`

**File:** `models/components.py`

```python
class ValveComponent(BaseComponent):
    # ... existing fields ...

    def calculate_head_loss(
        self,
        flow_gpm: float,
        velocity_fps: float,
        fluid_props: "FluidProperties",
    ) -> float:
        """Calculate valve head loss.

        Uses Cv if available, otherwise falls back to K-factor
        based on valve type and position.
        """
        if self.cv is not None:
            # Cv-based calculation: Q = Cv * sqrt(dP / SG)
            # Rearranged: dP = SG * (Q / Cv)^2
            # Convert dP (psi) to head (ft): h = dP / (0.433 * SG)
            sg = fluid_props.specific_gravity
            dp_psi = sg * (flow_gpm / self.cv) ** 2
            return dp_psi / (0.433 * sg)
        else:
            # K-factor based calculation
            from ..services.solver.k_factors import get_valve_k_factor
            k = get_valve_k_factor(self.valve_type, self.position)
            g = 32.174  # ft/s²
            velocity_head = velocity_fps ** 2 / (2 * g)
            return k * velocity_head
```

### 3. Implement for `HeatExchanger`

**File:** `models/components.py`

```python
class HeatExchanger(BaseComponent):
    # ... existing fields ...

    def calculate_head_loss(
        self,
        flow_gpm: float,
        velocity_fps: float,
        fluid_props: "FluidProperties",
    ) -> float:
        """Calculate heat exchanger head loss.

        Scales quadratically from design conditions:
        h_loss = design_drop * (Q / Q_design)^2
        """
        if flow_gpm <= 0:
            return 0.0

        # pressure_drop is in project units (psi), convert to feet
        sg = fluid_props.specific_gravity
        design_head_loss_ft = self.pressure_drop / (0.433 * sg)

        # Quadratic scaling
        flow_ratio = flow_gpm / self.design_flow
        return design_head_loss_ft * (flow_ratio ** 2)
```

### 4. Implement for `Strainer`

**File:** `models/components.py`

```python
class Strainer(BaseComponent):
    # ... existing fields ...

    def calculate_head_loss(
        self,
        flow_gpm: float,
        velocity_fps: float,
        fluid_props: "FluidProperties",
    ) -> float:
        """Calculate strainer head loss.

        Uses K-factor if available, otherwise scales from design
        pressure drop.
        """
        if self.k_factor is not None:
            # K-factor based
            g = 32.174  # ft/s²
            velocity_head = velocity_fps ** 2 / (2 * g)
            return self.k_factor * velocity_head
        elif self.pressure_drop is not None and self.design_flow is not None:
            # Fixed pressure drop with quadratic scaling
            sg = fluid_props.specific_gravity
            design_head_loss_ft = self.pressure_drop / (0.433 * sg)
            flow_ratio = flow_gpm / self.design_flow if self.design_flow > 0 else 0
            return design_head_loss_ft * (flow_ratio ** 2)
        else:
            return 0.0
```

### 5. Implement for `Orifice`

**File:** `models/components.py`

```python
class Orifice(BaseComponent):
    # ... existing fields ...

    def calculate_head_loss(
        self,
        flow_gpm: float,
        velocity_fps: float,
        fluid_props: "FluidProperties",
    ) -> float:
        """Calculate orifice head loss.

        Uses discharge coefficient and orifice diameter.
        Q = Cd * A * sqrt(2 * g * h)
        Rearranged: h = (Q / (Cd * A))^2 / (2 * g)
        """
        import math

        if flow_gpm <= 0:
            return 0.0

        # Convert flow to ft³/s
        flow_cfs = flow_gpm / 448.831

        # Orifice area in ft²
        orifice_area_ft2 = math.pi * (self.orifice_diameter / 24) ** 2  # diameter in inches

        g = 32.174  # ft/s²
        cd = self.discharge_coefficient

        # h = (Q / (Cd * A))^2 / (2 * g)
        head_loss = (flow_cfs / (cd * orifice_area_ft2)) ** 2 / (2 * g)
        return head_loss
```

### 6. Update solver to use Protocol

**File:** `services/solver/network.py`

```python
from ...protocols import HeadLossCalculator

def calculate_component_head_loss(
    component: Component,
    flow_gpm: float,
    velocity_fps: float,
    fluid_props: FluidProperties,
) -> float:
    """Calculate head loss for any component.

    Args:
        component: Network component
        flow_gpm: Flow rate in GPM
        velocity_fps: Velocity in ft/s
        fluid_props: Fluid properties

    Returns:
        Head loss in feet, or 0.0 if component doesn't cause loss
    """
    if isinstance(component, HeadLossCalculator):
        return component.calculate_head_loss(flow_gpm, velocity_fps, fluid_props)
    return 0.0
```

### 7. Add helper for valve K-factors

**File:** `services/solver/k_factors.py`

```python
def get_valve_k_factor(
    valve_type: ValveType,
    position: float | None = None,
) -> float:
    """Get K-factor for valve based on type and position.

    Args:
        valve_type: Type of valve
        position: Valve position (0=closed, 1=open), None for full open

    Returns:
        K-factor for head loss calculation
    """
    # Base K-factors for fully open valves (from Crane TP-410)
    base_k = {
        ValveType.GATE: 0.2,
        ValveType.BALL: 0.05,
        ValveType.BUTTERFLY: 0.3,
        ValveType.GLOBE: 10.0,
        ValveType.CHECK: 2.5,
        ValveType.STOP_CHECK: 13.0,
    }

    k = base_k.get(valve_type, 1.0)

    # Adjust for position (simplified model)
    if position is not None and position < 1.0:
        # K increases as valve closes (simplified exponential model)
        # At 50% open, K is roughly 4x the full-open value
        if position <= 0:
            return float('inf')  # Closed valve
        k = k / (position ** 2)

    return k
```

## Acceptance Criteria

- [ ] `HeadLossCalculator` Protocol defined in `protocols/components.py`
- [ ] `ValveComponent.calculate_head_loss()` implemented
- [ ] `HeatExchanger.calculate_head_loss()` implemented
- [ ] `Strainer.calculate_head_loss()` implemented
- [ ] `Orifice.calculate_head_loss()` implemented
- [ ] `get_valve_k_factor()` helper added to `k_factors.py`
- [ ] Solver updated to use Protocol for component losses
- [ ] All existing tests pass
- [ ] New unit tests for each `calculate_head_loss()` implementation
- [ ] mypy passes with Protocol type checking

## Testing

```python
import pytest
from opensolve_pipe.models.components import ValveComponent, ValveType, HeatExchanger
from opensolve_pipe.protocols import HeadLossCalculator
from opensolve_pipe.services.fluids import get_water_properties


@pytest.fixture
def water_props():
    return get_water_properties(68.0, "F")


def test_valve_satisfies_protocol():
    valve = ValveComponent(
        id="valve-1",
        name="Test Valve",
        elevation=0.0,
        valve_type=ValveType.GATE,
        cv=100.0,
    )
    assert isinstance(valve, HeadLossCalculator)


def test_valve_cv_head_loss(water_props):
    valve = ValveComponent(
        id="valve-1",
        name="Test Valve",
        elevation=0.0,
        valve_type=ValveType.GATE,
        cv=100.0,
    )
    # At 100 GPM through Cv=100, dP = SG * (Q/Cv)² = 1.0 * 1.0 = 1.0 psi
    # Head = 1.0 / 0.433 = 2.31 ft
    h_loss = valve.calculate_head_loss(100.0, 5.0, water_props)
    assert h_loss == pytest.approx(2.31, rel=0.01)


def test_heat_exchanger_quadratic_scaling(water_props):
    hx = HeatExchanger(
        id="hx-1",
        name="Test HX",
        elevation=0.0,
        pressure_drop=5.0,  # psi at design
        design_flow=100.0,  # GPM
    )
    # At design flow: h = 5.0 / 0.433 = 11.55 ft
    h_at_design = hx.calculate_head_loss(100.0, 5.0, water_props)
    assert h_at_design == pytest.approx(11.55, rel=0.01)

    # At half flow: h = 11.55 * 0.25 = 2.89 ft
    h_at_half = hx.calculate_head_loss(50.0, 2.5, water_props)
    assert h_at_half == pytest.approx(2.89, rel=0.01)


def test_strainer_k_factor(water_props):
    strainer = Strainer(
        id="str-1",
        name="Test Strainer",
        elevation=0.0,
        k_factor=2.0,
    )
    # velocity_head = v² / (2g) = 10² / 64.35 = 1.55 ft
    # h_loss = K * velocity_head = 2.0 * 1.55 = 3.1 ft
    h_loss = strainer.calculate_head_loss(100.0, 10.0, water_props)
    assert h_loss == pytest.approx(3.1, rel=0.01)
```

## Dependencies

- PR #120 

### Circular Import Prevention

The `calculate_head_loss` methods may need to import from `services.solver.k_factors`. Use local imports inside methods to avoid circular dependencies:

```python
def calculate_head_loss(self, ...):
    from ..services.solver.k_factors import get_valve_k_factor
    # ...
```
